### PR TITLE
Tint only the docs image the selection actually covers

### LINKED ISSUE #870: Copying an image selected by clicking it puts nothing on the clipboard
### What happened:
Selecting an image by **clicking** it — the state where the resize handles
appear — and pressing Cmd+C writes **nothing** to the system clipboard. The
clipboard keeps whatever was copied before, so the next paste replays that
instead. It looks like "the image will not paste", but the copy never happened.

https://github.com/wafflebase/wafflebase/issues/726#issuecomment-5312821680

### What you expected to happen:

Copying an image that is selected puts it on the clipboard, and pasting it back
into a document inserts the image — the same as copying any other content.

### How to reproduce it (as minimally and precisely as possible):

1. Copy some text (anything) so the clipboard has a known previous value.
2. Open a docs document and insert an image via the toolbar.
3. **Click** the image so it is selected — the resize handles appear.
4. Press Cmd+C.
5. Put the caret somewhere in the paragraph and press Cmd+V.

→ The text from step 1 is pasted.

### Environment:

- Operating system: macOS
- Browser and version: Whale 4.39.410.14 (arm64)

### PR DESCRIPTION
**Pull request by shjohw12**
> **Rescoped 2026-08-30.** This PR opened as three commits fixing #870 (copying a
> click-selected image). #984 landed on `main` in the meantime and fixes #870
> through a different design — `TextEditor.imageSelectionProvider` /
> `imageDeleteHandler` — so commits 1 and 3 here are redundant with it and have
> been dropped. The **selection-tint** commit is not on `main` and the defect it
> fixes is still live, so the branch is rebased onto `main` carrying that alone.
> The five review-panel findings that applied only to the dropped commits
> (`readOnly` gate, `selectImageAt` guard, the table-cell copy carve-out, the
> header/footer promotion, the sibling `getBlock` sites) went with the code they
> were about — those call paths no longer exist on this branch.

## Summary

Selecting one of several adjacent images tinted its **neighbour** as well, from
roughly the third image onwards.

https://github.com/user-attachments/assets/f4c1ca89-e1fb-4c62-9046-b0c09ef35632

An image is opaque and hides the selection highlight painted beneath it, so
`DocCanvas` fills the picture again with the selection colour whenever it
intersects the selection. The fill rounds its left/top edge and `renderRun`
rounds identically, so the tint lands exactly on the picture — that part was
right. The intersection **test** reused that rounded value, while selection
rectangles arrive from `computeSelectionRects` unrounded: it compared a
screen-space edge against a layout-space one.

Neighbours share an edge exactly, so half a pixel is the whole margin — a
selection over one *touches* the other, and rounding up makes the touch an
overlap:

```
image 2 layout [259.9 → 381.8]     image 3 starts at 381.8 — touching
old test       round(259.9) + 121.9 = 381.9   →  381.8 < 381.9  →  false overlap
```

Pasted pictures are scaled to the content width, so widths are fractional and
the first rounding that goes up lands around the third image — which is why the
first two looked fine. The rule moved into `imageIntersectsSelection` and
compares unrounded. **The fill's output is unchanged**; only the decision to
fill is.

### Table cells carried their own copy of the same comparison

`renderTableContent` paints cell content itself — `DocCanvas`'s run loop never
reaches it — so `table-renderer.ts` held a second, independently-rounded copy of
the overlap test and cell images kept the half-pixel over-tint. It now calls the
shared helper.

That is also why the helper lives in `image-selection-overlay.ts` rather than
next to either caller: `doc-canvas` imports `table-renderer`, so
`table-renderer` importing `imageIntersectsSelection` from `doc-canvas` would be
a cycle. `image-selection-overlay` already owns `ImageRect` and is imported by
both.

`@wafflebase/core/geometry`'s `rectsIntersect` is deliberately not reused: it
counts a shared edge as overlapping (Google Slides lasso semantics), and a
shared edge is exactly the case this has to answer *no* to.

### The regression is asserted through `DocCanvas`, not through a copy of it

The first version of the test verified the extracted helper plus a hand-copied
duplicate of the render loop. `imageIntersectsSelection` is a plain,
unconditional rectangle test — it cannot express the bug, because the bug was in
what the *call site* passed it. A test that reimplements the loop stays green if
`Math.round` comes back at the one place it can.

The test now builds a real layout, runs `DocCanvas.render` against a recording
2D context, and reads the fills it emitted. Restoring `Math.round(box.x/ box.y)`
at `doc-canvas.ts` fails it:

```
× image 3 of 4 tints only itself   AssertionError: expected [ 1, 2 ] to deeply equal [ 2 ]
× image 4 of 4 tints only itself   AssertionError: expected [ 2, 3 ] to deeply equal [ 3 ]
```

— which is exactly the reported symptom, "from roughly the third image onwards".
The cell path is pinned the same way against `renderTableContent`, and fails
2 of 4 when its own rounded comparison is restored.

Both the highlight and the tint fill with `Theme.selectionColor`; the test tells
them apart by order (the highlight pass runs once per page before any run is
drawn) and *asserts* that the first `rects.length` fills equal the selection
rectangles, so the split is a checked fact rather than an assumption.

## Linked Issues

Fixes #\
(#870 is fixed on `main` by #984, not by this branch — see the note above.)

## Author checklist

- [x] I searched existing issues and PRs and confirmed this is not a duplicate.
- [x] I read every line of this diff and can explain why each change is necessary.
- [x] Changes follow the conventions in the touched packages; any deviations are explained in *Why* above.
- [x] If AI tools assisted with this PR, I noted where in *Notes for Reviewers* below.

## Verification

- [ ] verify-self — green in the checks list
- [ ] verify-integration — green in the checks list (or explicit skip reason below)

Skip reason (if applicable): none — left unticked until CI reports. `pnpm verify:fast`
and the pre-push `verify:self` harness (22/32 lanes, 10 filtered, 0 failed) are
green locally on this base (`origin/main` @ `51243ec43`).

`packages/docs/test/view/image-selection-tint.test.ts` — 14 tests:

- 6 unit tests on `imageIntersectsSelection`, including both shared-edge cases
  and the vertical one the earlier revision left vacuous.
- 4 tests driving `DocCanvas.render` over four adjacent fractional-width images,
  asserting each tints only itself. The width is `121.9` on purpose: an integer
  width does not reproduce it, because the row starts on a whole pixel and it
  takes an accumulating fraction to push a neighbour across a rounding boundary.
- 4 tests driving `renderTableContent` over the same row inside a table cell.

Each group was confirmed to fail with its own rounding restored, then to pass
with it removed.

## Risk Assessment

- **User-facing risk:** low. No painted pixel changes — only the decision to
  paint. The body path's fill arguments are byte-identical to before; the cell
  path's are too.
- **Data/security risk:** none. Rendering only; no persistence, network or CRDT
  schema.
- **Rollback plan:** single commit, revert it.

## Notes for Reviewers

**AI assistance:** the original three commits were written with Claude Code. The
rescope, the `table-renderer` fix and the rewritten test in this revision were
also done with Claude Code; the failure output quoted above was produced by
actually restoring each rounding and running the suite.

**Follow-up work (unchanged from the original filing, none of it on this branch):**

- Copying an image puts no image on the clipboard for *other* apps — `text/plain`
  is the bare `U+FFFC`, so pasting into Slack or Word yields a box glyph. The
  real fix needs `navigator.clipboard.write()` with a `ClipboardItem`
  (`packages/slides` has the pattern), which turns `handleCopy` asynchronous.
- Pasting over a selected image inserts before it instead of replacing it.
- Windows Shift+Delete / Ctrl+Insert are not wired to cut/copy.
- A copied image's selection band is line-height, not image-height.

### CODE REVIEW

File: packages/docs/src/view/editor.ts [hunk 1]
```diff
@@ -2248,7 +2248,23 @@ export function initialize(
...
-      const key = e.key;
+      // A remote edit ahead of the image moves it off the stored offset.
+      // Everything that paints already resolves the offset and finds no
+      // image, so the handles vanish on their own — but the offset itself
+      // stayed live, and every branch below acts on it: cut took out
+      // whichever character had moved into its place, Delete removed it.
+      // Nothing on screen said a keystroke was about to do anything.
+      // Drop the selection to match what is drawn, and let the key reach
+      // the text path.
+      const selectedBlock = doc.getBlock(selectedImage.blockId);
```
[the remark this answers is not in the document]
Comment (shjohw12) [reply packages/docs/src/view/editor.ts]: Good catch — confirmed and applied in f3cdfe2e0.

`getBlock` does throw (`document.ts:186`), so the `!selectedBlock` half of the guard was unreachable: a peer removing the block outright would take the keydown listener down with it rather than clearing the selection. Reproduced before fixing — deleting the block and pressing a key raised `Error: Block not found: b1`.

Switched to `findBlock` and pinned it with a regression test (`a remote delete of the whole block does not throw`, mac × win), which fails with the throw on the previous revision.

One observation while checking: the same dead `getBlock(...)` + falsy-check pattern sits at four pre-existing sites in this file (lines ~3172, ~3190, ~3663, ~3677), including `getSelectedImage`, which resolves the same `selectedImage.blockId` this guard does. Left them alone as out of scope for this PR — happy to open a follow-up if you would like them swept.

**Comment by hackerwins** [comment]
@claude review this.

**Comment by hackerwins** [comment]
Went through the review panel's 7 blocking findings. Result: **2 fixed, 5 no longer apply** — the branch has been rebased and rescoped, and the title/body are rewritten to match.

### Why the rescope

#984 landed on `main` while this was open and fixes #870 — the same issue — through a different design (`TextEditor.imageSelectionProvider` / `imageDeleteHandler`, with the key handler consuming Cmd/Ctrl+C without `preventDefault`). Commits 1 (`Copy a selected image…`) and 3 (`Drop a stale image selection…`) are redundant with it, so they are dropped.

The tint commit is **not** on `main`, and I confirmed the defect is still live there — `doc-canvas.ts` still does `const ix = Math.round(pageX + pl.x + run.x)` and compares that against the unrounded selection rects. So the branch is now `main` + that one commit.

### Fixed

| # | Finding | What changed |
|---|---|---|
| 7 | *Tint regression is never exercised through DocCanvas — the test re-implements the render loop* | Correct, and the important one. `imageIntersectsSelection` is an unconditional rectangle test; it cannot express the bug, because the bug was in what the call site passed it. The test now builds a real layout and runs `DocCanvas.render` against a recording 2D context, reading the fills it emitted. Restoring `Math.round` at `doc-canvas.ts` now fails `image 3 of 4` and `image 4 of 4` — exactly the reported "from roughly the third image onwards". |
| — | *(Blast-radius suggestion)* `table-renderer.ts:439` still uses the rounded comparison | Fixed too, and pinned with 4 more tests driving `renderTableContent`. `renderTableContent` paints cell content itself, so `DocCanvas`'s run loop never reaches it — cell images kept the over-tint. This is why the helper moved to `image-selection-overlay.ts`: `doc-canvas` imports `table-renderer`, so importing it the other way would be a cycle. |
| — | *(Correctness suggestion)* the test box drops `pageY`, making the vertical assertion vacuous | Gone with the hand-copied loop; the render path supplies the real `pageY`. A vertical shared-edge unit test was added as well. |

### No longer applicable

Findings 1, 3, 5, 6 (`readOnly` gate on `imageKeyHandler`, `selectImageAt` guard, sibling `doc.getBlock` sites), 2 (header/footer promotion) and 4 (table-cell copy carve-out) were all about code paths this branch no longer contains — they belonged to the copy/stale-selection commits that #984 replaced. Worth checking whether they still hold against #984's implementation on `main`; if so they deserve their own issue rather than being fixed here, since none of that code is in this diff.

### On `rectsIntersect` reuse

Not adopted, deliberately, and it is now documented at the helper: `@wafflebase/core/geometry`'s `rectsIntersect` counts a shared edge as intersecting (Slides lasso semantics), and a shared edge is precisely the case this has to answer *no* to. Adding a strict variant to the shared package for one docs rendering rule looked like the wrong trade; happy to revisit if a third caller shows up.

`pnpm verify:fast` and the pre-push `verify:self` harness (22/32 lanes, 10 filtered, 0 failed) are green on `origin/main` @ `51243ec43`.
